modules/ingest-common/src/main/java/org/opensearch/ingest/common/ScriptProcessor.java

```diff
@@ -102,8 +102,11 @@ public IngestDocument execute(IngestDocument document) {
         } else {
             ingestScript = precompiledIngestScript;
         }
-        ingestScript.execute(document.getSourceAndMetadata());
-        CollectionUtils.ensureNoSelfReferences(document.getSourceAndMetadata(), "ingest script");
+        IngestDocument mutableDocument = new IngestDocument(document);
+        ingestScript.execute(mutableDocument.getSourceAndMetadata());
+        CollectionUtils.ensureNoSelfReferences(mutableDocument.getSourceAndMetadata(), "ingest script");
+        document.getSourceAndMetadata().clear();
+        document.getSourceAndMetadata().putAll(mutableDocument.getSourceAndMetadata());
```

Comment:
What exactly is happening here? A naive reading of this code looks like `document` is being mutated but `mutableDocument` is not being mutated.

Reply:
`mutableDocument` is being mutated and then after validation changes are being applied to original doc.